### PR TITLE
Admin report for domains which should have data source restriction FF enabled

### PR DESCRIPTION
**Pull request by Charl1996**
## Product Description
A new admin report is added to allow us to identify domains which have large amounts of data (forms or cases) which might cause data source rebuilds to be very slow. We use a heuristic (number of cases/form > 1,000,000) to determine whether or not those domains should have the `RESTRICT_DATA_SOURCE_REBUILD` feature flag enabled. 

The report looks like this:

image

The column, `UCR rebuild restriction status` (feel free to suggest other names), shows the user whether the domain
1. is currently being restricted from rebuilding data sources (i.e. have the FF enabled)
2. is currently **not** being restricted from rebuilding UCRs (i.e. does not the FF enabled)
3. should be restricted from rebuilding data sources (i.e. should have the FF enabled)
4. should not be restricted (i.e. should have the FF disabled; although I don't think well necessarily have a case like this)

When the FF should be enabled or disabled, the column will have a link (see image above) to the FF page.

In addition, a weekly email is sent to a particular admin email (soltech AEs) using this report's data to identify domains which we should look at. The email links to this report.

## Technical Summary
[Ticket](https://dimagi.atlassian.net/browse/SC-3662)

I decided to add a separate class, `UCRRebuildRestrictionTable`, handling the table feeding the `UCRDataLoadReport` report. This was decided so that we can easily have access to the report data when sending the email (e.g. specifying in the email the number of identified domains based on a certain filter) without having to create a mock request object as is needed when initializing a report class.  

## Feature Flag
No particular feature flag

### Safety story
Tested locally. Added some tests

### Automated test coverage
Added automated tests

### QA Plan
No QA from QA team, but will have AE's test it.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/hqwebapp/tasks.py
Comment (Charl1996) [inline corehq/apps/hqwebapp/tasks.py:271]: Now that I review this, this should probably be an environment variable.

File: corehq/apps/hqadmin/reports.py
Comment (mkangia) [inline corehq/apps/hqadmin/reports.py:349]: you can select only one domain?

File: corehq/apps/hqadmin/reports.py
Comment (mkangia) [inline corehq/apps/hqadmin/reports.py:454]: Can we memoize `RESTRICT_DATA_SOURCE_REBUILD.get_enabled_domains()` so we do this only once?

File: corehq/apps/hqwebapp/tasks.py
Comment (mkangia) [reply corehq/apps/hqwebapp/tasks.py:271]: Do we already have this email? if not, might as well use the soltech email we used to report issues to soltech only.

File: corehq/apps/hqwebapp/tasks.py
Comment (mkangia) [inline corehq/apps/hqwebapp/tasks.py:279]: does not reverse already handle this for you? Is the function `get_url_base` doing what you need?

**Review by mkangia — COMMENTED** [review]
Nice PR @Charl1996 , well drafted.

Just wanted to check about pagination on this report. I see we have added 4 options for filter status that we are using to choose to show domains. Is that to work with the pagination or the pagination is not supported on this report?

If you are going to use pagination, Just noting that any filter on the report should be applied when fetching the records to show, then paginated and then they should not be reduced.

A simple code improvement could be do order methods in order of use, the [Newspaper metaphor of clean code](https://www.baeldung.com/cs/clean-code-formatting#vertical-formatting).

File: corehq/apps/hqadmin/reports.py
Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:349]: As of now, yes.

File: corehq/apps/hqwebapp/tasks.py
Comment (Charl1996) [reply corehq/apps/hqwebapp/tasks.py:279]: `get_url_base` does exactly what I need! Thanks!

File: corehq/apps/hqadmin/reports.py
Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:454]: Good point

File: corehq/apps/hqadmin/reports.py
Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:367]: Cool, will look into this a bit.

**Comment by Charl1996** [comment]
@mkangia 

> Do we already have this email? if not, might as well use the soltech email we used to report issues to soltech only.

I see that goes to `solutions-feedback@dimagi.com` 

https://github.com/dimagi/commcare-hq/blob/f838c3e98520d8cee6e68db413c49e6ce26b6d95/corehq/apps/hqwebapp/views.py#L890

As I don't think it would make sense to send this report to the feedback email I think it's probably fine as is?

**Comment by Charl1996** [comment]
Please note that I force pushed due to lots of small updates baked into existing commits.

File: corehq/apps/hqadmin/reports.py
Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py]: I had a chat with AE and the domain filter is redundant.

**Comment by mkangia** [comment]
> As I don't think it would make sense to send this report to the feedback email I think it's probably fine as is?

Oh ya, feedback email should not be used.
Though, instead of `solutions-tech-app-engineers@dimagi.com` we could use `soltech-app-engineers@dimagi.com`.

**Comment by mkangia** [comment]
@Charl1996 please let me know when its ready for review again. I believe you would either respond to the pagination query I had or update code accordingly.

**Comment by Charl1996** [comment]
@mkangia 

> @Charl1996 please let me know when its ready for review again. I believe you would either respond to the pagination query I had or update code accordingly.

Ah, I forgot to push the "re-request review" button. It's ready for review. Thanks for checking in.

File: corehq/apps/hqadmin/reports.py
Comment (ajeety4) [inline corehq/apps/hqadmin/reports.py]: nit: rename to `restrict_rebuild_column_data`

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py]: Actually, now that I think about it, how does `_ucr_rebuild_restriction_status_column_data` sound? It's more in line with the column name.

File: corehq/apps/hqadmin/reports.py
Comment (mkangia) [inline corehq/apps/hqadmin/reports.py:501]: does this mean that we are creating all rows on each page and then picking the ones for the page?
Don't think we should be doing that.

Comment (mkangia) [reply corehq/apps/hqadmin/reports.py:501]: Honestly, I don't think we need pagination on this report since we are only fetching domains with UCR feature flag enabled which should not be many.

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:501]: @mkangia 

> does this mean that we are creating all rows on each page and then picking the ones for the page?

Correct. I do feel slightly uncomfortable about that, but I'm not sure how else I could achieve pagination, considering that I need to pull the domains first and then check if the domain applies based on the filter value. In this sense I need to construct all the applicable rows, but then paginate for the UI.

Note: I did check with an AE on whether the filters are necessary in the first place (since they seem kinda nice-to-have) but it was confirmed that they would be very helpful.

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:501]: > I don't think we need pagination on this report since we are only fetching domains with UCR feature flag enabled which should not be many.

The domains on prod [having this FF](https://www.commcarehq.org/hq/flags/edit/user_reports/) is already quite long.

File: corehq/apps/hqadmin/reports.py
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py]: @Charl1996 Could you please update the screenshot in the PR description to show the updated report filters since the domain filter has now been removed?

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py]: Yes! Let me do that now

File: corehq/apps/hqadmin/reports.py
Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:501]: Just to be clear on what's happening here:
1. A list of domains are fetched which has the UCR feature flag enabled
2. The domains are iterated and a subset of domains are selected based on the filters (this is something that needs to happen on the application level as there's various things to check)
3. This reduced (might be little, might be a lot) list of domains are then only paginated.

Considering this, if we get rid of pagination there won't be any performance increase except the time it takes to slice a list in python, as pagination can only happens as a last step for this feature (not ideal, but don't see a way around it).

If we get rid of the filters the UX is deteriorated significantly. 

The trade-off is thus the performance on HQ vs time spent by AE's scrolling down a list trying to determine if some domains need a feature flag enabled.

I'm leaning towards trading off performance on HQ, as its not a significant enough concern in my opinion. To give an idea of why I'm thinking this way: check out performance [on staging](https://staging.commcarehq.org/hq/admin/ucr_data_load/) for this report. Production has just a little more than double the amount of records.

Comment (mkangia) [reply corehq/apps/hqadmin/reports.py:501]: one small correction here / correct me if I am wrong:
> The domains are iterated and a subset of domains are selected based on the filters (this is something that needs to happen on the application level as there's various things to check)

The domains are iterated and data (case and form count) is fetched for all of them one at a time.
For each domain, It is after the data fetch, we decide if we want to list the domain or not based on filters.
And then once we have the final list of valid domains based on filters, we then paginate.

I agree that performance implication isn't noticeable on staging and then possibly would be same on production.
But I would not consider that as a validation to go with this approach.

If you look at the code, it is like, if these were SQL records, you are fetching all records from DB and then paginating using a list, instead of asking SQL to give you a set of records. And its reasonable because the table is small?
So, I believe, irrespective of the performance implication, you wont go with the above approach to paginate the list instead of via SQL.

The program is fetching and processing all domains for each page anyway, so if we just skip the pagination, the page itself won't become any fast for any single page, but the user won't do multiple requests (different pages) anymore.
So, if you wish to keep the filters, we should skip pagination. Initially, we noticed that the number of UCRs is high, but then the program is going through all of them anyway even now.

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:501]: > The domains are iterated and data (case and form count) is fetched for all of them one at a time.

This is not true anymore. It was initially the way I did it by changed it [here](https://github.com/dimagi/commcare-hq/pull/34813/commits/76f9c74842c0f8ddabeb2c4bc766acb05a5bffb5). Now all the domains' case data is fetched in one query to ES and the forms data is fetched in another query to ES.

 > The program is fetching and processing all domains for each page anyway, so if we just skip the pagination, the page itself won't become any fast for any single page, but the user won't do multiple requests (different pages) anymore.

This is true. I see your point. Yeah, so if we skip pagination and keep the filters I can now see how that would be an improvement to the current approach. I'll update

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py]: Reverted pagination [here](https://github.com/dimagi/commcare-hq/pull/34813/commits/7860d98bdcd80f51798849bd110a2ae94334ffdb).

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py]: I noticed that the page still looks paginated (maybe this is the standard look) but all records are loaded in the table, since there's no data loading going on when I switch between pages.

File: corehq/apps/hqadmin/reports.py
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py]: Thanks!

File: corehq/apps/hqadmin/reports.py
Comment (zandre-eng) [inline corehq/apps/hqadmin/reports.py:383]: Would it be possible for us to make this `memoized`? Then we only need to retrieve the list of UCR domains once.

Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:383]: Good idea

**Comment by zandre-eng** [comment]
@Charl1996 Thinking through this report, I'm wondering if there would be any value in having an action column as well? This could be even only be enabled for relevant rows that have a "Should Disable" or "Should Enable". Clicking on the button in the new column would then make an Ajax call to the back-end and add/remove the domain from the feature flag. 

This feels more like a nice-to-have, but it could make things a bit easier by not having to go and manually add/remove domains under the feature flags page.

**Comment by Charl1996** [comment]
@zandre-eng Currently I do something like that, but not as explicitly as an action column. What currently happens is that when the status is "should disable" or "should enable" a link to the FF edit page is included for quick access. See [here](https://github.com/dimagi/commcare-hq/blob/bc15ccf3d4a7dcb8b504afdca291f19a9f57eee9/corehq/apps/hqadmin/reports.py#L451).

**Comment by zandre-eng** [comment]
> Currently I do something like that, but not as explicitly as an action column. What currently happens is that when the status is "should disable" or "should enable" a link to the FF edit page is included for quick access. See [here](https://github.com/dimagi/commcare-hq/blob/bc15ccf3d4a7dcb8b504afdca291f19a9f57eee9/corehq/apps/hqadmin/reports.py#L451).

@Charl1996 Ah, that's perfect. Thanks for the explanation.

**Review by zandre-eng — APPROVED** [review]

File: corehq/apps/hqadmin/reports.py
Comment (mkangia) [reply corehq/apps/hqadmin/reports.py]: Do you need to set `disable_pagination=True` on the class?

**Comment by mkangia** [comment]
> a link to the FF edit page is included for quick access

I like that approach.

**Review by mkangia — APPROVED** [review]
Good to go once we are able to remove the pagination entirely.

**Comment by Charl1996** [comment]
@mkangia 
> Good to go once we are able to remove the pagination entirely.

You mean from the UI?

**Comment by mkangia** [comment]
> You mean from the UI?

@Charl1996 yes. If there is no pagination there should be no pagination on UI.

> I've been trying to do this by setting ajax_pagination and disable_pagination settings on the report, which hides the pagination spec [here](https://github.com/dimagi/commcare-hq/blob/d83800db8704ee412672c5104ce68ca12c27b8e6/corehq/apps/reports/generic.py#L1073-L1074), however, it doesn't seem to work as the pagination still shows up.

Spent a couple of mins and I could not get this to work either. let me know if you get blocked on this, I will give this a better try.

**Comment by Charl1996** [comment]
@mkangia I'm having no success so far with removing the pagination, but I'm wondering if we should not complete that part in another PR so as to unblock this one. Thoughts?

**Comment by mkangia** [comment]
Hey @Charl1996 

Got behind this this time, and looks like what we need to set is `use_datatables = False` to get rid of pagination.

I tried `show_all_rows`, `disable_pagination` and `ajax_pagination` first but pagination still showed up. Even though, new request was not going if i switched to page 2. Not sure why that was? I had 16 records and page size was 10. 
I guess it still loads all data and pagination is not via ajax. Not sure where that decision happens.

**Comment by Charl1996** [comment]
@mkangia 

> looks like what we need to set is use_datatables = False to get rid of pagination

Nice! Thanks!

File: corehq/apps/hqwebapp/tasks.py
Comment (mkangia) [inline corehq/apps/hqwebapp/tasks.py:262]: did the day_of_week=1 sent the email at a wrong time?

File: corehq/apps/hqadmin/reports.py
Comment (mkangia) [inline corehq/apps/hqadmin/reports.py:483]: do we need all these? I thought `use_datatables` was all that was needed?

File: corehq/apps/hqwebapp/tasks.py
Comment (Charl1996) [reply corehq/apps/hqwebapp/tasks.py:262]: It seems like it. Apparently there was emails sent every hour or so, but this new configuration seems to have fixed it. It's weird.

File: corehq/apps/hqadmin/reports.py
Comment (Charl1996) [reply corehq/apps/hqadmin/reports.py:483]: @mkangia 

I don't think they're technically necessary, but I thought of keeping it there to make it obvious to the developer that disabling pagination is the goal, since `user_datatables` doesn't really give me an indication that I'm trying to get rid of pagination. I also thought it's maybe a good idea to be explicit by setting the values to their current state.

Let me know if you think I should remove it or if it's fine to keep it.

Comment (mkangia) [reply corehq/apps/hqadmin/reports.py:483]: Okay :+1: 

I am fine with leaving it how you have done it.

I believe the other options make pagination static, i.e all records are loaded on the UI but there is still pagination shown so there is no ajax request to fetch more records but the user still sees pages. I could see that behavior locally.

> I thought of keeping it there to make it obvious to the developer that disabling pagination is the goal

I like that. :+1:

**Review by mkangia — APPROVED** [review]

**Pull request merged**
